在go版本>=1.20时generic mock失败

# env

* wsl2 Arch Linux on win11
* go version：1.21.1 & 1.20.8 & 1.19.8 & 1.18.8


# problem

使用go1.21.1和go1.20.8编译(编译命令禁止了内联和优化`go build -gcflags '-N -l' -o generic-go1.21.1`)运行examples中的generic时报错，使用go1.18.8和go1.19.8编译后执行可以正常输出结果为2。

error：

```
panic: Can not find CALL instruction

goroutine 1 [running]:
github.com/go-kiss/monkey.getFirstCallFunc(0x487be0)
        /home/ruo/go/src/monkey/monkey_amd64.go:118 +0x277
github.com/go-kiss/monkey.patchValue({0x48fd60?, 0x4a9cc8?, 0x1?}, {0x48fe20?, 0x4a9cd0?, 0x48f300?}, 0xc0000a6000)
        /home/ruo/go/src/monkey/monkey.go:145 +0x58a
github.com/go-kiss/monkey.Patch({0x48fd60?, 0x4a9cc8?}, {0x48fe20?, 0x4a9cd0?}, {0xc00004a6f0, 0x1, 0x7fb6b70ee5b8?})
        /home/ruo/go/src/monkey/monkey.go:54 +0x19e
main.main()
        /home/ruo/go/src/monkey/examples/generic/generic.go:26 +0xb2
```

# generic.go代码

```go
package main

import (
	"fmt"

	"github.com/go-kiss/monkey"
)

type S1[T int | float64] struct {
	i T
}

func (s *S1[T]) Get() T {
	return s.i
}

type S1__monkey__[T int | float64] struct {
	S1[T]
}

func (s *S1__monkey__[T]) Get() T {
	return s.i + 1
}

func main() {
	monkey.Patch((*S1[int]).Get, (*S1__monkey__[int]).Get, monkey.OptGeneric)
	s := S1[int]{i: 1}
	fmt.Println(s.Get()) // display 2
}
```


# 问题分析

明确了问题出现的场景及看了`getFirstCallFunc`的实现后，我用gdb对go1.18.8编译出的generic-go1.18.8和go1.21.1编译出的generic-go1.21.1进行了调试

go1.18.8反编译如下所示：


![image](https://github.com/go-kiss/monkey/assets/23385044/2768f604-fdd6-44f2-8a6d-dc66238a0cf3)


go1.21.1反编译如下所示：
![image](https://github.com/go-kiss/monkey/assets/23385044/0bd61524-31f7-4e26-8f68-89deb4cbcdc8)

而在`getFirstCallFunc`中的判断逻辑是要求call指令的上一条指令使用RAX寄存器
```go
if i.Op == x86asm.CALL && lastLea.Args[0].(x86asm.Reg) == x86asm.RAX
```

所以问题应该出现在这里，我将这里的判断逻辑修改为使用RAX或RBX寄存器后可以正常执行了，但是还不确定go1.20进行了什么改动导致的这一问题

--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -5,38 +5,38 @@ jobs:
     name: Test on Linux
     runs-on: ubuntu-latest
     steps:
-    - name: Set up Go 1.18
+    - name: Set up Go 1.20
       uses: actions/setup-go@v1
       with:
-        go-version: 1.18
+        go-version: 1.20
       id: go
     - name: Check out code into the Go module directory
       uses: actions/checkout@v1
     - name: Test
-      run: go test -gcflags=-l
+      run: go test -gcflags='all=-N -l'
   test-macos:
     name: Test on Mac
     runs-on: macos-latest
     steps:
-    - name: Set up Go 1.18
+    - name: Set up Go 1.20
       uses: actions/setup-go@v1
       with:
-        go-version: 1.18
+        go-version: 1.20
       id: go
     - name: Check out code into the Go module directory
       uses: actions/checkout@v1
     - name: Test
-      run: go test -gcflags=-l
+      run: go test -gcflags='all=-N -l'
   test-windows:
     name: Test on Windows
     runs-on: windows-latest
     steps:
-    - name: Set up Go 1.18
+    - name: Set up Go 1.20
       uses: actions/setup-go@v1
       with:
-        go-version: 1.18
+        go-version: 1.20
       id: go
     - name: Check out code into the Go module directory
       uses: actions/checkout@v1
     - name: Test
-      run: go test -gcflags=-l
+      run: go test -gcflags='all=-N -l'

--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -1,42 +1,18 @@
 name: Go
 on: [pull_request]
 jobs:
-  test-linux:
-    name: Test on Linux
-    runs-on: ubuntu-latest
+  test:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+        version: [1.18, 1.19, "1.20", 1.21]
+    runs-on: ${{ matrix.os }}
     steps:
-    - name: Set up Go 1.20
-      uses: actions/setup-go@v1
-      with:
-        go-version: 1.20
-      id: go
-    - name: Check out code into the Go module directory
-      uses: actions/checkout@v1
-    - name: Test
-      run: go test -gcflags='all=-N -l'
-  test-macos:
-    name: Test on Mac
-    runs-on: macos-latest
-    steps:
-    - name: Set up Go 1.20
-      uses: actions/setup-go@v1
-      with:
-        go-version: 1.20
-      id: go
-    - name: Check out code into the Go module directory
-      uses: actions/checkout@v1
-    - name: Test
-      run: go test -gcflags='all=-N -l'
-  test-windows:
-    name: Test on Windows
-    runs-on: windows-latest
-    steps:
-    - name: Set up Go 1.20
-      uses: actions/setup-go@v1
-      with:
-        go-version: 1.20
-      id: go
-    - name: Check out code into the Go module directory
-      uses: actions/checkout@v1
-    - name: Test
-      run: go test -gcflags='all=-N -l'
+      - name: Set up Go ${{ matrix.version }}
+        uses: actions/setup-go@v1
+        with:
+          go-version: ${{ matrix.version }}
+      - name: Check out code into the Go module directory
+        uses: actions/checkout@v1
+      - name: Test
+        run: go test -gcflags='all=-N -l'

--- a/README.md
+++ b/README.md
@@ -14,8 +14,6 @@ Go 语言猴子补丁（monkey patching）框架。
 
 Bouke 已经不再维护原项目，所以只能开一个新项目了🤣。
 
-有兴趣的同学也可以加微信 `taoshu-in` 讨论，拉你进群。
-
 ## 快速入门
 
 首先，引入 monkey 包
@@ -62,7 +60,9 @@ func main() {
 
 ## 注意事项
 
-1. Monkey 需要关闭 Go 语言的内联优化才能生效，比如测试的时候需要：`go test -gcflags=all=-l`。
+1. Monkey 需要关闭 Go 语言的内联优化才能生效，比如测试的时候需要：`go test -gcflags='all=-N -l'`。
 2. Monkey 需要在运行的时候修改内存代码段，因而无法在一些对安全性要求比较高的系统上工作。
 3. Monkey 不应该用于生产系统，但用来 mock 测试代码还是没有问题的。
-4. Monkey 目前仅支持 amd64 指令架构，支持 linux/macos/windows 平台。
+4. Monkey 目前仅支持 amd64 指令架构，支持 linux/macos/~~windows~~[^win] 平台。
+
+[^win]: 目前在 Win 平台下 Go 1.20 版本之后会报错。欢迎熟悉 Win 平台的同学提供 PR。问题解决之前建议 Win 用户使用 WSL。

--- a/monkey_amd64.go
+++ b/monkey_amd64.go
@@ -1,6 +1,9 @@
 package monkey
 
 import (
+	"runtime"
+	"strings"
+
 	"golang.org/x/arch/x86/x86asm"
 )
 
@@ -95,13 +98,12 @@ func getFirstCallFunc(from uintptr) uintptr {
 	f := rawMemoryAccess(from, 1024)
 
 	s := 0
-	var lastLea x86asm.Inst
 	for {
 		i, err := x86asm.Decode(f[s:], 64)
 		if err != nil {
 			panic(err)
 		}
-		if i.Op == x86asm.CALL && lastLea.Args[0].(x86asm.Reg) == x86asm.RAX {
+		if i.Op == x86asm.CALL {
 			arg := i.Args[0]
 			imm := arg.(x86asm.Rel)
 			next := from + uintptr(s+i.Len)
@@ -111,15 +113,15 @@ func getFirstCallFunc(from uintptr) uintptr {
 			} else {
 				to = next - uintptr(-imm)
 			}
-			return to
+			f := runtime.FuncForPC(to)
+			// 泛型函数的名字中包含 [...]
+			if strings.Index(f.Name(), "[") > 0 {
+				return to
+			}
 		}
 		s += i.Len
 		if s >= 1024 {
 			panic("Can not find CALL instruction")
-		}
-		// 有些情况下会生成 NOPW 指令，需要跳过
-		if i.Op == x86asm.LEA {
-			lastLea = i
 		}
 	}
 }

